fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccount.java

```diff
@@ -1573,21 +1576,36 @@ public void validateAccountBalanceDoesNotBecomeNegative(final String transaction
                         throw new PlatformApiDataValidationException(dataValidationErrors);
                     }
                 }
-
             }
             lastSavingsDate = transaction.transactionLocalDate();
-
         }
 
-        BigDecimal withdrawalFee = null;
-        BigDecimal transactionAmount = null;
         if (isOverdraft()) {
             if (runningBalance.minus(minRequiredBalance).isLessThanZero()) {
-                throw new InsufficientAccountBalanceException("transactionAmount", getAccountBalance(), withdrawalFee, transactionAmount);
+                overdraftLimitExceededException("transactionAmount", getAccountBalance(), transactionAmount);
+            }
+        }
+    }
+
+    public void validateAccountBalanceDoesNotViolateOverdraft(final List<SavingsAccountTransaction> savingsAccountTransaction,
+            final BigDecimal amountPaid) {
+        if (savingsAccountTransaction != null) {
+            SavingsAccountTransaction savingsAccountTransactionFirst = savingsAccountTransaction.get(0);
```

Comment:
@rrpawar96 I was wondering whether the actual accountBalance of the Savings account would not be better candidate here instead of the running balance of the last transaction?

I dont think we should fetch all of the transactions to use only the first...

Reply:
@adamsaghy , I do agree with this as this uses lots of memory space. I have seen challenges with accounts that have lots of transactions. However, the account balance is prone to inaccuracy. This means that if we are using it at this point, there must be another process class the accurately updates the account balance. 

@rrpawar96  if you remember the [issue](https://issues.apache.org/jira/browse/FINERACT-1147)

Reply:
@francisguchie thanks for highlighting this issue. I think if the account balance is not accurate  at particular scenarios that is a major issue and should be addressed with priority.
However if the fix is not straightforward, i can imagine a workaround where not all of the transactions are fetched, but only the very last (if the transaction running balance is more reliable than the account balance). That way the performance and memory impact would be less.

what do you think?

Reply:
Yes, the JPA doesn't support the 'Limit 1' keyword in the Query to fetch a single record, So, the alternative that I will be using now is Pageable Param, is there any other alternative you think would fix this part? or Pageable would be fine? @adamsaghy

Reply:
@rrpawar96 You can use Pageable, or something like this:
Optional<Passenger> findFirstBySavingsAccountAndReversedFalseAndDateOfEqualsOrderByIdDesc();

However I have changed a little bit the query... 
Provided transaction date is equals or less... do we really need the "less" part?
Order by date is not enough as multiple transaction could happen on that day. Use Id as the secondary order by param to get the last transaction.

However i was wondering what is the purpose of this query. Shall it return the very last transaction of the savings account which is not reversed? In that case we dont need the transaction date and just order by id Desc.

Please help me to understand better this query...

Reply:
@rrpawar96 and @francisguchie Using the account balance is really that problematic? I thought the account balance is automatically recalculated for every payment activity (transaction).

Reply:
@adamsaghy , I think its not problematic to use the account balance

Reply:
@francisguchie In that case i would rather rely on the account balance instead of trying to figure out which transaction was the very last and what is the running balance of it. I mean both could work and i believe both of them should be always the same (hopefully).

Reply:
@adamsaghy, The account balance history on the basis of the past dates is saved in only one table i.e. "m_savings_account_transaction (SavingsAccountTransaction)" in the "running account balance" field and there is no field as "account balance" in SavingsAccountTransaction (m_savings_account_transaction table). Although the "account balance" field is in another table "m_savings_account" that keeps the update of the latest balance of the account and not of the past date (history)

Reply:
@rrpawar96 Yeah... i should have read more carefully the story about this change :)
Seems you're right, you need to find the very last transaction of that particular day (pay charge transaction date) and based on the running balance of that transaction decide whether it is violating the overdraft rule or not.

To do that, I would recommend this:
`Optional<SavingsAccount> findFirstBySavingsAccountAndReversedFalseAndDateOfLessThanEqualOrderByDateOfDescIdDesc();`

But this one is pretty complex, so maybe Pagination would be just fine as well:
`

@Query("select sat from SavingsAccountTransaction sat where sat.savingsAccount.id = :savingsId and sat.dateOf <= :transactionDate and sat.reversed=false ORDER BY sat.dateOf DESC, sat.id DESC")
     List<SavingsAccountTransaction> retrieveRunningBalanceBySavingsAccountIdAndDate(@Param("savingsId") Long savingsId,
             @Param("transactionDate") LocalDate transactionDate, PageRequest.of(0, 1));`

Reply:
@adamsaghy thanks